### DISCUSSION TITLE
Accessing Camera in Home crashes Homebridge with ERR_INVALID_ARG_TYPE

### DISCUSSION THREAD

**Discussion by fancypantsifier** [homebridge-simplisafe3/homebridge-simplisafe3#280] [Q&A] [answered]
Everything seems to be working ok except the cameras. If I click on one in the Home app to see a feed, I immediately see this in my Homebridge status log:

```
[1/11/2022, 1:51:48 PM] TypeError [ERR_INVALID_ARG_TYPE]: The "file" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:329:5)
    at validateString (node:internal/validators:129:11)
    at normalizeSpawnArguments (node:child_process:432:3)
    at spawn (node:child_process:602:13)
    at SS3SimpliCam.handleStreamRequest (/usr/local/lib/node_modules/homebridge-simplisafe3/accessories/simplicam.js:473:46)
[1/11/2022, 1:51:48 PM] Got SIGTERM, shutting down Homebridge...
[1/11/2022, 1:51:48 PM] [Simplisafe Alarm] handleStreamRequest with request: { sessionID: 'ce28f198-5d25-445b-afd5-1eac48629682', type: 'stop' }
[1/11/2022, 1:51:53 PM] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
```
and the feed says "No Response" / "This camera is not responding."

Snapshots also don't work: I only ever see "No response" and a black feed.

I have two cameras and I see this for each when Homebridge starts (I have debugging enabled for the Simplisafe plugin):
```
  cameraStatus: {
    firmwareVersion: '2.6.2.126',
    netConfigVersion: '',
    camAgentVersion: '',
    lastLogin: 1641913275,
    lastLogout: 1641913212,
    wlanMac: <REDACTED>
    fwDownloadVersion: '',
    fwDownloadPercentage: 0,
    recovered: false,
    recoveredFromVersion: '',
    batteryPercentage: 0,
    _id: <REDACTED>
    initErrors: [],
    setupTimestamp: 1638593297,
    speedTestTokenCreated: 1638593297
  },
```
Homebridge is running on a Mac Mini. One thing that just occurred to me is that this is an old Mac Mini, running an old OS (10.11.6) and version node.js version v15.14.0. Perhaps that's related?

**Comment by fancypantsifier** [comment]
Update, I've now manually installed some version updates:
```
Node.js Version: v16.13.2
Npm Version: v8.1.2
```
For _one_ of my cameras, when I bring up its room in the Home app, I get this in the Homebridge log:
```
[1/11/2022, 2:40:17 PM] [Simplisafe Alarm] Handling camera snapshot for 'Front Door' at 640x360
[1/11/2022, 2:40:21 PM] [Simplisafe Alarm] Closed 'Front Door' snapshot request with 23kB image
[1/11/2022, 2:41:28 PM] [Simplisafe Alarm] Handling camera snapshot for 'Front Door' at 1280x720
[1/11/2022, 2:41:32 PM] [Simplisafe Alarm] Closed 'Front Door' snapshot request with 20kB image
[1/11/2022, 2:41:42 PM] [Simplisafe Alarm] Handling camera snapshot for 'Front Door' at 1280x720
[1/11/2022, 2:41:45 PM] [Simplisafe Alarm] Closed 'Front Door' snapshot request with 20kB image
[1/11/2022, 2:41:55 PM] [Simplisafe Alarm] Handling camera snapshot for 'Front Door' at 1280x720
[1/11/2022, 2:41:58 PM] [Simplisafe Alarm] Closed 'Front Door' snapshot request with 20kB image
```
But I still just see black images in the display.
For the other camera, I see a blue dot at upper right and "now" at lower right in the Home app, but nothing the Homebridge log and nothing but black images. 
Note that I do have the privacy shutter enabled while the alarm is off (which it is now).

When I click on the camera to open a live view, I get now get a slightly different error:
```
[37m[1/11/2022, 2:45:48 PM] [39m[31mTypeError: The "file" argument must be of type string. Received undefined[39m
[31m    at new NodeError (node:internal/errors:371:5)[39m
[31m    at validateString (node:internal/validators:119:11)[39m
[31m    at normalizeSpawnArguments (node:child_process:503:3)[39m
[31m    at spawn (node:child_process:691:13)[39m
[31m    at SS3SimpliCam.handleStreamRequest (/usr/local/lib/node_modules/homebridge-simplisafe3/accessories/simplicam.js:473:46)[39m
[37m[1/11/2022, 2:45:48 PM] [39mGot SIGTERM, shutting down Homebridge...
```

**Reply by fancypantsifier** [reply]
I just reinstalled homebridge on top of itself to pick up the node/npm updates. Do I also need to reinstall the simplisafe3 plugin?

**Reply by fancypantsifier** [reply]
Also just tried using [ffmpeg-for-homebridge](https://github.com/homebridge/ffmpeg-for-homebridge) and setting the custom ffmpeg path in the Simpliesafe3 plugin to use it. I still don't see snapshots, but when I try to open the camera for streaming, homebridge no longer crashes. Now I get:
```
[1/11/2022, 3:25:50 PM] [Simplisafe Alarm] FFMPEG received first frame
[1/11/2022, 3:25:50 PM] [Simplisafe Alarm] dyld: Library not loaded: /usr/local/opt/xz/lib/liblzma.5.dylib
  Referenced from: /usr/local/bin/ffmpeg
  Reason: Incompatible library version: ffmpeg requires version 8.0.0 or later, but liblzma.5.dylib provides version 6.0.0

[1/11/2022, 3:25:50 PM] [Simplisafe Alarm] Camera stopped streaming
```
I'm trying to see if/how I can update liblzma on my old Mac mini.

**Comment by shamoon** [comment, marked as answer]
Hi @fancypantsifier I believe this is a duplicate of #141 , see https://github.com/homebridge-simplisafe3/homebridge-simplisafe3/issues/141#issuecomment-662102274

tldr, ffmpeg is not compatible with your MacOS, not much I can do there, sorry

**Reply by fancypantsifier** [reply]
Ah yes, thanks, that does seem like almost the same thing. I will see if I can custom build my own ffmpeg on my ancient Mac. 

The one difference I see is that the issue you linked seems not to have trouble with the snapshots. Those aren't working for me either, and that seems independent of ffmpeg. Any suggestions there?

**Reply by shamoon** [reply]
I’m a little confused by your previous comment but if you have privacy shutter closed the plug-in will respect that and not open it for snapshots. Does it work when alarm is armed? Can you check on multiple devices (ideally including a Mac as there’s an iOS 15.2 bug that prevents snapshots updating for a lot of users)

**Reply by fancypantsifier** [reply]
Thanks, I was misunderstanding how the Privacy shutter works. When the Privacy shutter is disabled, snapshots for the cameras are mostly working (see below), albeit with the warning that "The image snapshot handler for the given accessory is slow to respond!"

Interestingly, _one_ of the camera's snapshots still aren't working on my phone, which is running iOS 15.1, not 15.2. Snapshots work for both cameras on my Macbook, however. I don't even see any log messages indicating an attempt to get snapshots for that camera when I open the Home app on my phone, but on the Macbook, I see the same logging for both cameras. I appreciate all the help here. If you have any ideas on what would prevent snapshots from working for just one of the two cameras on my phone, that'd be great, but it doesn't really seem like this is this plugin.

I'm also currently building my own ffmpeg-for-homebridge. We'll see how that goes.

**Reply by shamoon** [reply]
Perhaps snapshots could be bug I mentioned https://discussions.apple.com/thread/253512399 https://www.reddit.com/r/HomeKit/comments/rpwcww/camera_snapshots_stuck_showing_now_but_image_is/ (many other places if you google it)

**Reply by fancypantsifier** [reply]
Ok, didn't realize the snapshot thing was so widespread. I'll investigate elsewhere. Thanks for the pointer.

I finally got ffmpeg built, but am now seeing these errors. I'm not sure if this suggests something wrong with my build or if it could be something set up wrong with the Simplisafe plugin. Ever seen anything like this? (Error writing trailer / Operation not permitted)

```
[1/12/2022, 12:27:18 PM] [Simplisafe Alarm] Start streaming video for camera 'Front Door'
[1/12/2022, 12:27:18 PM] [Simplisafe Alarm] /usr/local/bin/ffmpeg -re -headers Authorization: Bearer <REDACTED>/flv?x=1280&audioEncoding=AAC -map 0:0 -vcodec libx264 -tune zerolatency -preset superfast -pix_fmt yuv420p -r 20 -f rawvideo -vf scale=1280:-2 -b:v 284k -bufsize 568k -maxrate 284k -payload_type 99 -ssrc 5209053 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params QdPw9T7GaKdfqo/HiOqCC1/cWP6vNFgnuQzO/biq srtp://192.168.1.44:53597?rtcpport=53597&localrtcpport=53597&pkt_size=1378 -map 0:1 -acodec libfdk_aac -flags +global_header -profile:a aac_eld -ac 1 -ar 16k -b:a 24k -bufsize 48k -payload_type 110 -ssrc 11426258 -f rtp -srtp_out_suite AES_CM_128_HMAC_SHA1_80 -srtp_out_params sW1zV49wp2SayN7f5Ilc0UrTCI6TTcN72GyOQSj+ srtp://192.168.1.44:57188?rtcpport=57188&localrtcpport=57188&pkt_size=188
[1/12/2022, 12:27:18 PM] [Simplisafe Alarm] FFMPEG received first frame
[1/12/2022, 12:27:18 PM] [Simplisafe Alarm] ffmpeg version 4.3.1 Copyright (c) 2000-2020 the FFmpeg developers
  built with Apple LLVM version 8.0.0 (clang-800.0.42.1)
  configuration: --enable-videotoolbox --pkgconfigdir=/Users/<REDACTED>/dev/ffmpeg-for-homebridge-master/workspace/lib/pkgconfig --prefix=/Users/<REDACTED>/dev/ffmpeg-for-homebridge-master/workspace --pkg-config-flags=--static --extra-cflags=-I/Users/<REDACTED>/dev/ffmpeg-for-homebridge-master/workspace/include --extra-ldflags='-L/Users/<REDACTED>/dev/ffmpeg-for-homebridge-master/workspace/lib -L/opt/local/lib' --extra-libs='-lpthread -lm -lz -liconv' --enable-static --disable-debug --disable-shared --disable-ffplay --disable-doc --enable-openssl --enable-gpl --enable-version3 --enable-nonfree --enable-pthreads --enable-libvpx --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libx264 --enable-runtime-cpudetect --enable-libfdk-aac --enable-avfilter --enable-libopencore_amrwb --enable-libopencore_amrnb --enable-filters --enable-decoder=h264 --enable-network --enable-protocol=tcp --enable-libspeex --enable-demuxer=rtsp

[1/12/2022, 12:27:18 PM] [Simplisafe Alarm]   libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100

[1/12/2022, 12:27:18 PM] [Simplisafe Alarm] [https @ 0x7fd9fbc17380] No trailing CRLF found in HTTP header. Adding it.

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] Input #0, flv, from 'https://<REDACTED>/v1/<REDACTED>/flv?x=1280&audioEncoding=AAC':
  Metadata:
    encoder         : Lavf56.40.101
  Duration: 
[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] 00:00:00.00, start: 0.000000, bitrate: N/A
    Stream #0:0: Video: h264 (Main), yuvj420p(pc, bt709, progressive), 1280x720 [SAR 1:1 DAR 16:9], 20 fps, 20 tbr, 1k tbn, 40 tbc
    Stream #0:1: Audio: aac (LC), 16000 Hz, mono, fltp, 96 kb/s

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> h264 (libx264))
  Stream #0:1 -> #1:0
[1/12/2022, 12:27:22 PM] [Simplisafe Alarm]  (aac (native) -> aac (libfdk_aac))
Press [q] to stop, [?] for help

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] Output #1, rtp, to 'srtp://192.168.1.44:57188?rtcpport=57188&localrtcpport=57188&pkt_size=188':
  Metadata:
    encoder         : Lavf58.45.100
    Stream #1:0: Audio: aac (libfdk_aac) (ELD), 16000 Hz, mono, s16, 24 kb/s
    Metadata:
      encoder         : Lavc58.91.100 libfdk_aac

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] [swscaler @ 0x7fd9fc14b600] deprecated pixel format used, make sure you did set range correctly
[libx264 @ 0x7fd9fc85f000] using SAR=1/1
[libx264 @ 0x7fd9fc85f000] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.1 Cache64

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] [libx264 @ 0x7fd9fc85f000] profile High, level 3.1, 4:2:0, 8-bit
Output #0, rtp, to 'srtp://192.168.1.44:53597?rtcpport=53597&localrtcpport=53597&pkt_size=1378':
  Metadata:
    encoder         : Lavf58.45.100
    Stream #0:0: Video: h264 (libx264), yuv420p, 1280x720 [SAR 1:1 DAR 16:9], q=-1--1, 284 kb/s, 20 fps, 90k tbn, 20 tbc
    Metadata:
      encoder         : Lavc58.91.100 libx264
    Side data:
      cpb: bitrate max/min/avg: 284000/0/284000 buffer size: 568000 vbv_delay: N/A

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] av_interleaved_write_frame(): Operation not permitted

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] Error writing trailer of srtp://192.168.1.44:53597?rtcpport=53597&localrtcpport=53597&pkt_size=1378: Operation not permitted
frame=    1 fps=0.0 q=40.0 Lsize=       2kB time=00:00:00.17 bitrate=  80.0kbits/s speed=0.805x    
video:2kB audio:1kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] [libx264 @ 0x7fd9fc85f000] frame I:1     Avg QP:33.03  size:  1712
[libx264 @ 0x7fd9fc85f000] mb I  I16..4: 99.0%  0.4%  0.6%

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] [libx264 @ 0x7fd9fc85f000] 8x8 transform intra:0.4%
[libx264 @ 0x7fd9fc85f000] coded y,uvDC,uvAC intra: 0.7% 0.6% 0.0%
[libx264 @ 0x7fd9fc85f000] i16 v,h,dc,p: 88%  7%  5%  0%
[libx264 @ 0x7fd9fc85f000] i8 v,h,dc,ddl,ddr,vr,hd,vl,hu: 12% 12% 57%  5%  0%  5%  2%  0%  5%
[libx264 @ 0x7fd9fc85f000] i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 37% 12% 22%  4%  4%  4%  9%  2%  8%
[libx264 @ 0x7fd9fc85f000] i8c dc,h,v,p: 96%  1%  3%  0%
[libx264 @ 0x7fd9fc85f000] kb/s:273.92

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] Conversion failed!

[1/12/2022, 12:27:22 PM] [Simplisafe Alarm] Error: FFmpeg exited with code 1
```

**Comment by shamoon** [comment]
Yea I think that’s why ffmpeg-for-homebridge doesn’t support older versions of macOS. It’s not really something I know much about, not really in the scope of this plug-in. Sorry.

**Reply by fancypantsifier** [reply]
Sure, I get it. Thanks for taking a look. I'll ping that project to see if anyone has any idea or the solution to all this is to move to newer computer (even a raspberry pi!)

**Reply by shamoon** [reply]
IMHO Pi is def your best bet, will use much less power too

**Reply by fancypantsifier** [reply]
Well I found the Operation not permitted problem above: My (headless) Mac Mini had a dialog box waiting for me to approve network connections for ffmpeg. 🤦  Once I clicked Allow, the stream from one camera started working. 

The other camera is still failing, now with:
```
https://<REDACTED_IP>/v1/<REDACTED_KEY>/flv?x=1280&audioEncoding=AAC: Invalid data found when processing input
```

Perhaps this is still just some weird ffmpeg thing, which I understand you can't support, but given that your experience trying to support streams from Simplisafe cameras, I wonder if you've got any ideas here. It's just so weird that it words fine for one camera and not the other. They are the same model, purchased at the same time....

**Reply by shamoon** [reply]
Glad you got it working. Never seen that error before, certainly odd if theyre the same cameras etc. unless there are obvious settings differences between them

If you turn on debug and watch the logs you can test the full ffmpeg command even on a different machine, the whole thing that starts with `ffmpeg -re ...` (usually need to add quotes around the auth header and long URLs. Might answer the question of whether its ffmpeg

**Reply by fancypantsifier** [reply]
Lowering the resolution to 720p using "-s 1280x720" seems to have made the streaming work for both cameras. Snapshots still don't work for the one. No errors. I assume that doesn't use ffmpeg at all? I've tried unplugging and replugging that camera to see if it was in a bad state, but no luck. So weird.

**Reply by shamoon** [reply]
Yes odd indeed. 

You are correct snapshots doesnt use ffmpeg it parses a jpg from `https://media.simplisafe.com/v1/${this.simplicam.cameraDetails.uuid}/mjpg?x=${request.width}&fr=1` including token authentication ("Bearer" authentication) using the `accessToken` which is saved in the auth file.

You could try the URL yourself e.g. ensure that `curl -H "Authorization: Bearer ACCESS_TOKEN" "https://media.simplisafe.com/v1/CAMERA_UUID/mjpg?x=1280&fr=1" > screenshot.jpg` works. (Btw, mjpg is actually a stream so you'll have to manually close that curl request after a few seconds)

**Reply by fancypantsifier** [reply]
Yep, it seems to work when I do it manually from a terminal! But not in the Home app.

**Reply by shamoon** [reply]
Really wondering if this is the iOS bug I mentioned. Supposedly that’s fixed in upcoming iOS release
